**Shows how the T-Test for cosine distances was obtained**


In [ ]:
!pip install bertopic sentence-transformers umap-learn hdbscan scikit-learn pandas matplotlib
from bertopic import BERTopic
import pandas as pd

In [ ]:
import torch
print("GPU Available:", torch.cuda.is_available())
device = "cuda" if torch.cuda.is_available() else "cpu"

data = pd.read_csv('highconf_label_data.csv')

cat = 2.0

data = data[data['label'] == cat]
data = data.dropna(subset=['body']).reset_index(drop=True)

data = data['body'].to_list()

GPU Available: True


In [ ]:
from hdbscan import HDBSCAN

hdbscan_model = HDBSCAN(min_cluster_size=20,  # min cluster size before it can become its own
                        min_samples=15,       # how many neighbors it has before its part of a clsuter
                        cluster_selection_epsilon=0.1,
                        metric="euclidean",
                        prediction_data=True)

In [ ]:
from sentence_transformers import SentenceTransformer
from sklearn.feature_extraction.text import CountVectorizer

vectorizer_model = CountVectorizer(ngram_range=(1, 2), stop_words="english")
embedding_model = SentenceTransformer("all-MiniLM-L6-v2", device=device)
topic_model = BERTopic(embedding_model=embedding_model,
                       hdbscan_model=hdbscan_model,
                       vectorizer_model=vectorizer_model,
                       verbose=True)
topics, probs = topic_model.fit_transform(data)

2025-04-06 19:28:22,201 - BERTopic - Embedding - Transforming documents to embeddings.


Batches:   0%|          | 0/336 [00:00<?, ?it/s]

2025-04-06 19:28:33,498 - BERTopic - Embedding - Completed ✓
2025-04-06 19:28:33,503 - BERTopic - Dimensionality - Fitting the dimensionality reduction algorithm
2025-04-06 19:28:49,171 - BERTopic - Dimensionality - Completed ✓
2025-04-06 19:28:49,172 - BERTopic - Cluster - Start clustering the reduced embeddings
2025-04-06 19:28:49,731 - BERTopic - Cluster - Completed ✓
2025-04-06 19:28:49,741 - BERTopic - Representation - Fine-tuning topics using representation models.
2025-04-06 19:28:51,093 - BERTopic - Representation - Completed ✓


In [ ]:
topic_model.get_topic_info()

,Topic,Count,Name,Representation,Representative_Docs
0,-1,4560,-1_life_human_child_abortion,"[life, human, child, abortion, baby, unborn, m...",[you should look at the net consequence and ho...
1,0,973,0_murder_innocent_killing_murder murder,"[murder, innocent, killing, murder murder, kil...","[its not murder., so murder it., it's not murd..."
2,1,640,1_babies_killing babies_killing_baby,"[babies, killing babies, killing, baby, kill b...","[who is killing babies?, who is killing babies..."
3,2,298,2_religion_god_religious_abortion,"[religion, god, religious, abortion, murder, b...","[the bible is pro-abortion, so abortion should..."
4,3,285,3_link_lol_thank_like,"[link, lol, thank, like, womp, yeah, good, com...","[link, link, link]"
...,...,...,...,...,...
57,56,23,56_dem_audience_city_dem city,"[dem, audience, city, dem city, protests, dem ...",[all of the reasons add up to why he was banne...
58,57,22,57_unborn_right_right murder_right kill,"[unborn, right, right murder, right kill, murd...","[the right to murder*** an unborn child., ever..."
59,58,22,58_roe_wade_roe wade_overturning roe,"[roe, wade, roe wade, overturning roe, overtur...",[these are all stepping stones to overturning ...
60,59,21,59_heartbeat_heart_baby heartbeat_babys,"[heartbeat, heart, baby heartbeat, babys, hear...",[more gymnastics. when a developing human ba...


In [ ]:
pl_embed = topic_model.topic_embeddings_


In [ ]:
data = pd.read_csv('highconf_label_data.csv')

cat = 1.0

data = data[data['label'] == cat]
data = data.dropna(subset=['body']).reset_index(drop=True)

data = data['body'].to_list()

In [ ]:
from sentence_transformers import SentenceTransformer
from sklearn.feature_extraction.text import CountVectorizer

vectorizer_model = CountVectorizer(ngram_range=(1, 2), stop_words="english")
embedding_model = SentenceTransformer("all-MiniLM-L6-v2", device=device)
topic_model = BERTopic(embedding_model=embedding_model,
                       hdbscan_model=hdbscan_model,
                       vectorizer_model=vectorizer_model,
                       verbose=True)
topics, probs = topic_model.fit_transform(data)

2025-04-06 19:28:54,908 - BERTopic - Embedding - Transforming documents to embeddings.


Batches:   0%|          | 0/508 [00:00<?, ?it/s]

2025-04-06 19:29:17,049 - BERTopic - Embedding - Completed ✓
2025-04-06 19:29:17,050 - BERTopic - Dimensionality - Fitting the dimensionality reduction algorithm
2025-04-06 19:29:30,690 - BERTopic - Dimensionality - Completed ✓
2025-04-06 19:29:30,691 - BERTopic - Cluster - Start clustering the reduced embeddings
2025-04-06 19:29:31,478 - BERTopic - Cluster - Completed ✓
2025-04-06 19:29:31,485 - BERTopic - Representation - Fine-tuning topics using representation models.
2025-04-06 19:29:34,064 - BERTopic - Representation - Completed ✓


In [ ]:
topic_model.get_topic_info()

,Topic,Count,Name,Representation,Representative_Docs
0,-1,7449,-1_abortion_women_life_fetus,"[abortion, women, life, fetus, woman, people, ...",[and i'm sure your sympathies will do much to ...
1,0,2178,0_vote_like_just_people,"[vote, like, just, people, democrats, court, s...",[which is why republicans play to win. they p...
2,1,493,1_republicans_gop_republican_women,"[republicans, gop, republican, women, abortion...",[women who don't care that the government want...
3,2,357,2_blood_donate_kidney_organ,"[blood, donate, kidney, organ, body, save, org...",[nothing has a right to use your body. even to...
4,3,239,3_religious_religion_abortion_people,"[religious, religion, abortion, people, bible,...",[thank you for your respectful reply. ive resp...
...,...,...,...,...,...
74,73,22,73_incubators_incubator_incubation_forced incu...,"[incubators, incubator, incubation, forced inc...",[huh? i'm pointing out how its an even more ri...
75,74,21,74_children_animals_kids_want,"[children, animals, kids, want, women children...",[this is just straight up population and wealt...
76,75,21,75_forced birth_forced_birth_forced pregnancy,"[forced birth, forced, birth, forced pregnancy...","[""forced birth"" isnt accurate because there's ..."
77,76,21,76_000_abortions_link_000 women,"[000, abortions, link, 000 women, year, births...","[so with respect to abortions, the first stati..."


In [ ]:
pc_embed = topic_model.topic_embeddings_

In [ ]:
import numpy as np
from scipy import stats
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
similarity_matrix = cosine_similarity(pl_embed)

distance_matrix = 1 - cosine_similarity(pl_embed)

from scipy.spatial.distance import squareform

pl_distances = squareform(distance_matrix, checks=False)

pl_distances

array([0.2624855 , 0.1350761 , 0.15209234, ..., 0.6033461 , 0.7009525 ,
       0.8185123 ], dtype=float32)

In [ ]:
similarity_matrixPc = cosine_similarity(pc_embed)

distance_matrixPc = 1 - cosine_similarity(pc_embed)

from scipy.spatial.distance import squareform

pc_distances = squareform(distance_matrixPc, checks=False)

In [ ]:
t_stat, p_val = stats.ttest_ind(pc_distances, pl_distances)

print(f"T-statistic: {t_stat:.4f}, P-value: {p_val:.8f}")

T-statistic: 4.9378, P-value: 0.00000082


### T-Test Results

- **First run**  
  - *T-statistic:* 2.17  
  - *P-value:* 0.0297  

- **Second run**  
  - *T-statistic:* 8.11  
  - *P-value:* < 0.00000001  

- **Third run**  
  - *T-statistic:* 4.94  
  - *P-value:* 0.00000082  


